Updated wallet to look up account names

tuxcanfly opened the pull request:
Added an index for mapping account names to account numbers. Updated some existing RPC calls to support accounts by looking up the account number and passing it to manager.

Working on updating the remaining few RPC calls.

davecgh committed:
Add conversion from legacy keystore if it exists.

davecgh committed:
Move keystore to legacy directory.

tuxcanfly committed:
Turn echo off when reading passwords

davecgh committed:
Switch to new waddrmgr package

This commit converts the wallet to use the new secure hierarchical
deterministic wallet address manager package as well as the walletdb
package.

The following is an overview of modified functionality:

- The wallet must now be created before starting the executable
- A new flag --create has been added to create the new wallet using wizard
  style question and answer prompts
- Starting the process without an existing wallet will instruct now
  display a message to run it with --create
- Providing the --create flag with an existing wallet will simply show an
  error and return

In addition the snacl package has been modified to return the memory after
performing scrypt operations to the OS.

Previously a runtime.GC was being invoked which forced it to release the
memory as far as the garbage collector is concerned, but the memory was
not released back to the OS immediatley.  This modification allows the
memory to be released immedately since it won't be needed again until the
next wallet unlock.

tuxcanfly commented:
@davecgh I was reading through bitcoind to match the accounts implementation for RPC calls. Looks like there are two ways accounts are handled:
- `CWalletTx` has a `strFromAccount` which is set by `sendfrom` and `sendmany`. Every input from such a tx is considered as belonging to `strFromAccount`. This is required because `sendfrom`, `sendmany` may use inputs from any account.
- Every tx output maps to an account using `mapAddressBook` which maps addresses to accounts.

The way I've currently implemented makes `sendfrom`, `sendmany` use only inputs from the given account (i.e. making `findEligibleOutputs` filter outputs by account), so we don't need to have the account field on tx. We can map both inputs and outputs to account the same way.

Is there any problem continuing this way? Or should we match bitcoind behaviour and tag the tx with the account it was sent from?

davecgh commented:
That sounds fine.  Accounts support in Bitcoin Core really doesn't work that well and they've discussed removing them altogether, so while keeping the API compatible is a good goal, we don't need to (and can't due to deterministic keys) match the behavior exactly.

davecgh commented:
Also, I know you're still working on this, but before we merge this, please squash the commits down into logical change groups.  For example, one that adds the necessary bits to waddrmgr, then another one that modifies the RPC calls.  The other piece like fixing doc and stuff should be a separate PR.

You'll also want to rebase this on top of the latest davec_intwaddrmgr branch as it has been updated quite a bit for new walletdb bits.

tuxcanfly commented:
Sure, will do that.

tuxcanfly commented:
Depends on https://github.com/conformal/btcjson/pull/32 for the account param change in `getbalance`, `listtransactions`

tuxcanfly committed:
Updated waddrmgr to manage account names

jrick committed:
Implement several btcd RPCs in wallet as well.

Wallet should handle these so that wallet clients don't end up
creating requests to btcd about the latest processed block, which is
not the same as wallet's most recently processed block.

By providing wallet clients with this info, we avoid a race where the
client thinks wallet has processed N blocks, but in fact is still
synced to N-1 (and perhaps currently processing transactions from
block N).  This can cause unexpected results for many of the
bitcoind-compatible RPC APIs due to their reliance on number of
confirmations, rather than using absolute block heights.

jrick committed:
Use the correct synced block when calculating confs.

This was previously using the most recently notified (by the chain
package) block, but transaction processing from this block may not be
finished yet.  Using this block's height to calculate the number of
confirmations is therefore incorrect, and can result in every RPC
handler missing transactions or returning transactions from the wrong
block.

adamkrellenstein commented:
It looks like Bitcoin Core will be getting rid of accounts? https://github.com/bitcoin/bitcoin/pull/5575

davecgh commented:
Aye, I've seen that.  We don't plan to get get rid of them however.  Rather, to make them work properly.

davecgh commented:
To expand on this, the reason is that BIP0032 HD key support is coming and it uses the BIP0044 account scheme.  So this will expose the ability to access those account branches.  This is different than the way Bitcoin Core does it where you only have individual private keys and you can actually move addresses between accounts and such.  Obviously, with BIP0044, you don't/can't move addresses between accounts.

adamkrellenstein commented:
@davecgh, ah, ok. That is rather different, and perfectly reasonable.

tuxcanfly committed:
Mark used addresses as such

jrick committed:
Consolidate and optimize zero functions.

This introduce a new internal package to deal with the explicit
clearing of data (such as private keys) in byte slices, byte arrays
(32 and 64-bytes long), and multi-precision "big" integers.

Benchmarks from a xeon e3 (Xor is the zeroing funcion which Bytes
replaces):

BenchmarkXor32  30000000                52.1 ns/op
BenchmarkXor64  20000000                91.5 ns/op
BenchmarkRange32        50000000                31.8 ns/op
BenchmarkRange64        30000000                49.5 ns/op
BenchmarkBytes32        200000000               10.1 ns/op
BenchmarkBytes64        100000000               15.4 ns/op
BenchmarkBytea32        1000000000               2.24 ns/op
BenchmarkBytea64        300000000                4.46 ns/op

Removes an XXX from the votingpool package.

davecgh committed:
Update btcwire path import paths to new location.

jrick committed:
Fix Go versions for Travis.

jrick committed:
Use Travis' container-based infrastructure.

jrick committed:
Address several issues pointed out by lint and vet.

This brings the entire tree closer to but not 100% goclean.sh clean.

davecgh committed:
Update btcnet path import paths to new location.

davecgh committed:
Update btcec path import paths to new location.

davecgh committed:
waddrmgr: Update documentation for walletdb.

This commit updates the documentation which discusses creating and opening
the manager to properly mention the wallet database namespace as well as
another typo.

davecgh committed:
waddrmgr: Update more documentation for walletdb.

davecgh committed:
Update btcjson path import paths to new location.

davecgh committed:
Update btcws path import paths to new location.

jrick committed:
Rescan after btcd reconnect.

Otherwise, if the websocket connection to btcd is lost and
reestablished, there is no sync to the current best block, and btcd
will not notify wallet of new relevant transactions.

jrick committed:
Use rescan for initial tx notification registrations.

Requires btcsuite/btcd#310.

jrick committed:
Fix build.

This fixes the build after btcsuite/btcd#302, which renamed the
txscript constant ScriptCanonicalSignatures constant to
ScriptVerifyDERSignatures.

dajohi committed:
Use txscript.StandardVerifyFlags

By using txscript.StandardVerifyFlags when creating and validating
transactions, we can ensure the transactions successfully created
won't be rejected due to script policy.

jrick committed:
Update README for recent changes.

jrick committed:
Prepare for release 0.5.0.

jrick committed:
Whitespace fix.

davecgh committed:
waddrmgr: Make create/update logic more explicit.

This commit makes the creation and updating of the address manager more
explicit so it's easier to upgrade in the future.

In particular, rather than treating the initial creation as an upgrade by
relying on creating the initial buckets on the fly on each load, the code
now explicitly provides distinct create and upgrade paths that are invoked
from the Create and Open functions, respectively.

It also adds some commented out sample code to illustrate how upgrades
should be done and a check to ensure bumping the version number without
writing upgrade code results in a new error, ErrUpgrade, being returned.

Finally, a test has been added for the new functionality.

davecgh committed:
Update golang.org/x/crypto import paths to new location.

Commit:
Add new --createtemp flag to btcwallet.

The createtemp flag is used to create a temporary simnet wallet for
use with btcsim or other testing suites.

The flag only works if the datadir and network to use it on (simnet)
are specified.

After starting btcwallet with the flag, the wallet immediately opens
and is functional.

tuxcanfly committed:
Remove early return on empty rescan

jrick committed:
Prepare for release 0.5.1.
